### PR TITLE
fixed #112, weather values should now always be correct way around

### LINKED ISSUE #112: Weather outputs flipped for degrees C and F
- `shell` bash
- `Operating system/version`: arch
- `Terminal name/version`: st
- `$TERM`

## Steps to Reproduce the Problem
Run tuxi with weather query: `tuxi weather 10001`

## Expected Behavior
32 ºF
0 ºC

## Actual Behavior
32 ºC
0 ºF

## The Problem
The degrees are flip flopped. I have committed the changes to my fork and will open a PR.

### PR DESCRIPTION
**Pull request by PureArtistry**

### CODE REVIEW

**Comment by BeyondMagic** [comment]
God of shell scripting

**Comment by PureArtistry** [comment]
lol, thanks

**Comment by sudocanttype** [comment]
We could just send this straight to main, because this is a high priority problem.

**Comment by sudocanttype** [comment]
Alright, I adapted the code so that it works ok on the main branch. I'm going to make a PR and commit that so that this bug finally gets fixed, and then we can figure out how to merge purert eventually to main.

**Comment by Bugswriter** [comment]
What the f* is this reliable. LOL

**Comment by sudocanttype** [comment]
I can test it on my machine. Not sure if there is a one/two liner way to do it, but maybe. Not super experienced in bash

**Comment by sudocanttype** [comment]
Ok, so for some reason my test script crashes when running your script. "(40/3)+4*6" specifically causes it to crash.
image

**Comment by PureArtistry** [comment]
so long as the temperature number in fahrenheit is always greater than it is in celsius then it's reliable. I don't know about a cleaner solution though, that's the best I could think of.


https://user-images.githubusercontent.com/53883649/108901671-109b4700-7613-11eb-924f-bfb91e281cc8.mp4

**Comment by PureArtistry** [comment]
> Ok, so for some reason my test script crashes when running your script. "(40/3)+4*6" specifically causes it to crash.
> image

that message pops up if a lock file is created on the previous run and not cleared, so something went wrong the time you run it before that.

in that example it found the rich answer (I think it''s that one) before the math answer.
the view all line is the answer it got first.

I'm going to fix that now

**Comment by PureArtistry** [comment]
sorted
scrot_20210223-202927

**Comment by Genghius** [comment]
seem good.

**Comment by BeyondMagic** [comment]
It's nice showing who added their scripts so we can blame them for bugs

91e6993ef1a8cebac91ac8bc1e0849d03b00c5ff for #91 added `Rich Cast` and `Simple Lists` btw

**Comment by BeyondMagic** [comment]
The option to pick answers to test is amazing, wow, this fixes #95

**Comment by PureArtistry** [comment]
@BeyondMagic thanks, the command line way is good for occasional checks but if you want to switch lots off when doing testing it's easier to just comment out those tests in the priority variable at the top of the script.

added your name btw

**Comment by PureArtistry** [comment]
I think I'm done adding stuff to this for now until it's gone through some testing and any bugs have been ironed out

**Comment by sudocanttype** [comment]
That's probably a good idea. I can help with the testing part

**Comment by sudocanttype** [comment]
@PureArtistry Can you run the test script and send the output? For some reason I can't run it on my machine. It freezes and crashes at random questions, even though it technically shouldnt
image

**Comment by PureArtistry** [comment]
the tests are crashing because of the lock file rapidly being deleted and recreated, because everything is forked when the main script exits there is still sometimes a request process running in the background that hasn't quite died yet (kill -9 isn't instant unfortunately, may need to add another lap or two to the loop)

the main script is exiting deletes the lock and I guess that gives one of the forks a chance to re-make it while the next run of the script starts and after it as done the check for the lock, because the lock as been made after the check as passed nothing on the new test can print out so it hangs

need to maybe add a sleep 0.1 to the testscript or I need to absolutely make sure everything is dead before exiting

**Comment by PureArtistry** [comment]
another alternative could be to make unique folder names each run using date strings, that would solve the issue but I think I still need to fiddle with the exit - cure the disease not the symptoms

**Comment by sudocanttype** [comment]
If it's easier, I could implement the .1 sec delay in the test file. It would most likely only be a temporary fix, as maybe someone in the future will try to run multiple tuxi commands in succession super fast, but I doubt it. It would be better if this got fixed on the tuxi file, but we can just call that problems for the future.

**Comment by PureArtistry** [comment]
give me half an hour and I can try a couple of things to see if they solve the problem, I know a couple of things to test, I was hoping I could be lazy and rely on the shell's wait function to tell the truth about stuff being finished.

**Comment by BeyondMagic** [comment]
Any way to disable the extra output after sending the process for /dev/null?

image

**Comment by sudocanttype** [comment]
Can't you direct the output to dev/null too?

**Comment by PureArtistry** [comment]
not sure why that's happening, those line numbers don't even match up with my script and the file path in that output is to a different version of the script I think (at least that's what it looks like from your prompt)

**Comment by BeyondMagic** [comment]
I just cloned right now to test the new commits, and I did build it (make). It sometimes outputs, sometimes not.

**Comment by PureArtistry** [comment]
what shell is your bin/sh? (readlink /bin/sh)

**Comment by BeyondMagic** [comment]
zsh, happens the same thing in bash. 
image

This doesn't support dash/posix, right?

**Comment by BeyondMagic** [comment]
bash, I edited the comment.

**Comment by PureArtistry** [comment]
thanks, looks like I am going to have to edit the exit, in bash it seems something works a little differently.

**Comment by PureArtistry** [comment]
what is your distro? it might be the version of xargs rather than the shell that's the problem

**Comment by BeyondMagic** [comment]
Manjaro 20.2.1 Nibia

image

**Comment by PureArtistry** [comment]
interesting, it's unlikely to be xargs then.

can you try adding a sleep 0.2 before the final exit and running it a couple of times please?

if it's what i think it might be that shouldn't happen again

**Comment by BeyondMagic** [comment]
Nop, still happening.

**Comment by PureArtistry** [comment]
I'm re-writing that bit of the script in a minute, need to write another kill loop rather than lazily using xargs to do it for me.

that command uses as many threads as the kernel will give it and forks off a bunch of shells to run the kill commands in parallel. Im guessing the script exits before all of those forks finish (not sure why though) then FD2 gets re connected to your shell and the errors (which were re-directed in the script) print out - maybe. also I may just be talking out of my arse

**Comment by PureArtistry** [comment]
@BeyondMagic - when you get chance could you test this new commit and see if it still spits out errors

**Comment by BeyondMagic** [comment]
Sorry for being late, I will test right now

**Comment by sudocanttype** [comment]
Not home rn :/ I'll test it as soon as i can.

**Comment by sudocanttype** [comment]
What country are you guys testing from?

**Comment by BeyondMagic** [comment]
:brazil:

**Comment by sudocanttype** [comment]
... strange. 
image
It outputted correctly, but froze with the same request a second later.

**Comment by PureArtistry** [comment]
Sorry I took a while to get back on this, needed sleep.

All of these problems are caused by the main tuxi script exiting before all the forks are dead, working on fixing that now
but @BeyondMagic's problem with the output is something of a mystery to me right now - I have never had that problem and can't re-create it. 

I'm going to work on the exit bit first as that *should* fix the problem but I don't understand why yet - going to need to figure that out

**Comment by BeyondMagic** [comment]
I'll try to fix it myself this problem since it seems that it's only happening with me.

**Comment by PureArtistry** [comment]
for some reason your terminal/shell is ignoring to redirect to dev/null, can't think of why right now.

do you get error output when you run tuxi -a search term?

**Comment by BeyondMagic** [comment]
Yeah, it's working fine with -a.

**Comment by PureArtistry** [comment]
can you try changing your shell to dash (or something that isn't using a fancy prompt) then run script
```
dash
./tuxi define error
```

sounds silly but I'm running low on ideas for why you are getting FD2 when you shouldn't be

**Comment by PureArtistry** [comment]
if you could also try editing line 532

```
kill -0 "$kids" 2>/dev/null && kill -9 "$kids" >/dev/null 2>&1
```

**Comment by BeyondMagic** [comment]
< can you try changing your shell to dash (or something that isn't using a fancy prompt) then run script

Fixed after changing first line to run dash » `#!/usr/bin/dash`

> if you could also try editing line 532

I''ll try it

**Comment by PureArtistry** [comment]
cool, thanks you're giving stuff to work on - seems this is a bash (or other bourne compatable) shell problem I'll have to look into

especially as most people's bin/sh is bash

**Comment by PureArtistry** [comment]
just did some testing myself and it's definitely a bash problem, looking into it now.

**Comment by BeyondMagic** [comment]
Yep, **zsh** and **dash** works now, I swear it wasn't working before your last commit.

**Comment by PureArtistry** [comment]
it was a slightly different problem before the last commit, although I only changed one line this version does something very different

**Comment by PureArtistry** [comment]
I'm close to fixing the problem, bash is mangling the priority variable somewhere (might be other bad outputs too)

stepping through the script now to find out where it happens, I'm guessing I'm using echo somewhere I shouldn't be because bash's echo is an annoying special snowflake that works differently to almost every other shell.

**Comment by BeyondMagic** [comment]
> ...because bash's echo is an annoying special snowflake that works differently to almost every other shell.

tl;dr gnu is bloat

image

**Comment by PureArtistry** [comment]
> isn't bash killing the processes automatically?

how do you mean?

**Comment by BeyondMagic** [comment]
> how do you mean?

nothing, I'm dumb

> edit2: yeah, changing 532 line to kill -0 "$kids" 1>/dev/null 2>&1 & resolved the problem for bash.

is there a problem with this? worked for dash and zsh as well.

**Comment by PureArtistry** [comment]
> edit2: yeah, changing 532 line to `kill -0 "$kids" 1>/dev/null 2>&1 &` resolved the problem for bash.

thanks for that tip, that will solve the output problem, I do still need to fix the hanging problem, working on that now - I know how to fix it, just means editing a lot of lines

**Comment by BeyondMagic** [comment]
> I do still need to fix the hanging problem, working on that now - I know how to fix it, just means editing a lot of lines

you can do it!!! I believe on you!! *motivation mode*

**Comment by PureArtistry** [comment]
I never learned bash so I'm not sure of all it's little quirks, I will add the 1 to the rest of the redirects while I'm going through the script

**Comment by PureArtistry** [comment]
Just need some other people to test this and see if there are any more bugs | [testoutputs.txt](https://github.com/Bugswriter/tuxi/files/6041424/testoutputs.txt)

**Comment by sudocanttype** [comment]
Aight i can test it

**Comment by Bugswriter** [comment]
Sorry I was busy with exam.
So is this finished like completely tested?
> will this fix our f & c swap. 
Should I merge ?

**Comment by sudocanttype** [comment]
I'm not sure. I am still running into the same bug as before, even after fetching from the new commit. (It freezes and crashes at random commands on the test script)

**Comment by BeyondMagic** [comment]
[dab892e](https://github.com/Bugswriter/tuxi/pull/144/commits/dab892e705f0f18fad523c526abe2ddb11b0c47c) fixed it for bash extra output
[66bdf00](https://github.com/Bugswriter/tuxi/pull/144/commits/66bdf000f3a209e58cd33b9dca5cd4dc8a1e5f7c) broke it

test script is working fine though, no freezes at all.

**Comment by PureArtistry** [comment]
sorry been a little busy.

I'll revert the xargs change again - it works on my machine but it seems to still cause trouble on others, it only saves a couple of ms anyway - not worth the trouble

I've nearly finished a new test script, will commit soon.

**Comment by BeyondMagic** [comment]
Sorry for being bringing you so many problems, but I need to ask, how fast is the main for your @PureArtistry? There's a huge difference for me.

image

There's no extra output because I put the fixed lines of `dab892e`.

**Comment by PureArtistry** [comment]
you don't need to time it like that use the -d flag and it will tell you how long it took to process.
doing it like that is going to give inconsistant results because the curl request takes a variable amount of time

**Comment by PureArtistry** [comment]
I'll try moving maths up in the priority and see if that will make a difference. the -b flag might also be faster

**Comment by BeyondMagic** [comment]
-d ?
image

**Comment by PureArtistry** [comment]
-d for debug or dev - it didn't put it in help because I didn't know if that was going to stay an option in the main version of the script or only for the dev version

those flags are in my version, the -b is in the help message for mine.

my version is definitely slower at some queries, the ones at the top of the priority for the main script but should be faster the further down the list you go.

**Comment by BeyondMagic** [comment]
anyway, the command time is comparing apples to apples there. I'll test the ones at the bottom of priority order

**Comment by PureArtistry** [comment]
also out of curiosity, what sort of storage is you $HOME on?

**Comment by PureArtistry** [comment]
please let me know if it is consistantly slower for all queries, it might just be worth ditching the multi-threading and relying on a better priority list - will solve issues, make the script simpler.

all the forking adds some extra time for the processing and the fact you have to repeatedly access storage on my version adds to execution time compared to the main version

**Comment by PureArtistry** [comment]
saying all that though, it never takes that long to run on mine
scrot_20210225-212047

maths is always a slower one, if you try quotes, weather, cast, define or lyrics with -b it's much faster
scrot_20210225-212719

**Comment by BeyondMagic** [comment]
Try that without the -d and -b flags.

**Comment by PureArtistry** [comment]
I'll finish this test script up as it will be a better way to test speed difference across all the different types of query.
the -b is so much faster because it's only checking for quotes, that is as fast as my version will go

is this running on an ssd? (what sort of drive is your $HOME on?)

**Comment by BeyondMagic** [comment]
> is this running on an ssd?
Nop.

**Comment by PureArtistry** [comment]
that is probably why mine is so much slower on your machine, it's looking like it might be a better idea to ditch the multi thread.
at best on my machine it's not much faster than the main and if your on something with a hdd it's going to be much slower

**Comment by BeyondMagic** [comment]
Yeah, my computer is from 2006, I can't do anything about it, but how much fast the main branch of Bugswriter is (using time command) compared to yours in your machine (using time command) as well?

**Comment by PureArtistry** [comment]
scrot_20210225-215348

**Comment by BeyondMagic** [comment]
Yeah, it's just priority order at all, using -t fixes it. Nevermind this, it's not a problem at all.

**Comment by PureArtistry** [comment]
out of interest could you test the -a flag, that only touches the hdd once - it will give a better indication of how much of a bottleneck it might be

**Comment by BeyondMagic** [comment]
image

**Comment by PureArtistry** [comment]
I meant test my version against itself with/without -a

nice to see mine win a round though (and sorry to be a pain btw, thanks for doing this)

**Comment by BeyondMagic** [comment]
my bad, my bad

image

**Comment by PureArtistry** [comment]
assuming that wasn't an abnormally long curl on the first test, the hdd thing is a problem

**Comment by PureArtistry** [comment]
I just had a brainwave and come up with a way around the lock file, I'm going to finish the tests first and then do some tweaking

**Comment by BeyondMagic** [comment]
I'll be waiting, for the next, 10 hours, I don't sleep anyway

**Comment by PureArtistry** [comment]
neither do I, hence why I have so much time for staring at vim

**Comment by BeyondMagic** [comment]
nice, love your whole contribution, seriously, you're really helping a lot.

**Comment by PureArtistry** [comment]
sorry I took a while, went off to get some food - test script is done
[testoutputs.txt](https://github.com/Bugswriter/tuxi/files/6047202/testoutputs.txt)

**Comment by BeyondMagic** [comment]
Everything is working fine now, no freezes for me, no extra output, @sudocanttype just need to test the rest.

**Comment by PureArtistry** [comment]
I'm still going to switch up the lock method now I've come up with a way to do it without having to access the filesystem.

a number of the test queries will need modifying, quite a few I haven't even tested in google, I just threw something in the file that I thought might work

**Comment by PureArtistry** [comment]
they are super easy to edit now though, you don't even have to touch the script itself

**Comment by sudocanttype** [comment]
image
wack. It returned quotes instead, and the terminal was stuck until I hit ctrl c

**Comment by PureArtistry** [comment]
I'm tweaking the main script now and getting rid of that lockfile and making it better respect priority
